zjit/src/hir.rs

```diff
@@ -1687,6 +1692,7 @@ impl Function {
             Insn::StringIntern { .. } => types::Symbol,
             Insn::StringConcat { .. } => types::StringExact,
             Insn::StringGetbyteFixnum { .. } => types::Fixnum.union(types::NilClass),
+            Insn::StringAppend { .. } => types::StringExact,
```

Comment:
Technically not stringexact because you could have `String << StringExact` in the current setup. Either make it no output (since it returns the lhs anyway) or have the output type `String`